### ISSUE TITLE
Add type annotations for mypy and jedi

### ISSUE THREAD

**Issue by barcharcraz** [conan-io/conan#4177]
Conan currently has quite a few attributes and methods which confuse jedi, mypy and a lot of other tooling. Adding annotations to these attributes can help packagers write conanfiles more quickly and ensure they haven't forgotten anything.

It's easiest to add annotations inline with the source code (using `# type:` comments) however they could also be defined in a standalone file. Additionally jedi can parse rst doc comments and extract types from those.

I've put together a pull request implementing type hint comments for the basic ConanFile fields.

**Comment by uilianries** [comment]
I think we could use [type hints](https://docs.python.org/3/library/typing.html) after to deprecate python 2 support.

**Comment by memsharded** [comment]
Hi @barcharcraz 

I would like to learn a bit better how type annotations help the users. Is that they are getting syntax highlight errors directly in their IDEs even before running conan? Would you have some captures, error messages, etc that you could share?

While I understand the use case for big codebases, and it is something that we might implement when moving to py3 only, I would like to see the value for conan users, that in general they manage just 1 file, and there are already lots of checks in the conan codebase to protect from wrong inputs.

**Comment by barcharcraz** [comment]
The largest thing is getting correct completions for conanfile members. Things like scm or cpp_info.libs should show up in Jedi.

For me, it makes it much easier to remember to add all the fields I need to the conanfile. I could amend the pull request to only add required dummy variables and not mess with typing or add type hints.

**Comment by lasote** [comment]
Yeah, as you said, I think the important thing is to autocomplete the "complex" objects, like the cppinfo, options, settings, default_options and so on. I think we should wait for Conan 2.0 where we will use python3 and use all the power from there.

**Comment by cassava** [comment]
I would like to add that having type hints in the main code repository would be enormously helpful to new contributors. Trying to figure out what type a particular variable *is* is a game of hide-and-seek currently...

**Comment by luka-papez** [comment]
Hi everyone

I can see that this has been planned for Conan 2.0 which is great news!
Being a big fan of both Conan and python type-hinting, I would gladly contribute somehow - be it implementation, review, whatever.

Are you aware of any work already underway regarding adding type hints to Conan 2.0? Besides @barcharcraz PR from 3 years ago, I couldn't find anything.

**Comment by memsharded** [comment]
The truth is that we have started doing something in the ``develop2`` branch, but we aren't going full typing. You can see there some classes like ``DataCache`` that start to use some type hints. Still we haven't found the right tooling or the right balance to add types everywhere, so we are not forcing it into Conan 2.0 codebase yet. We are super comfortable with the codebase as it is right now. 

One thing that I would like is runtime type checking, that would be something interesting, but as far as I know, this is not a standard thing yet? Is there any way we could do such type checking, at least during testing?

**Comment by memsharded** [comment]
I am closing this ticket by now:

- The vast majority of core maintainers are very comfortable without types
- This is a difficult to mark as "finished" task, as typing the whole codebase will take a lot of time, definitely nothing that will be ready for 2.0 (which is already in alpha stage)
- We will not oppose to add types to new code and incrementally add typing to Conan codebase. This is already happening naturally, specially in some interfaces (as the new cache, as I commented above). But we will not enforce typing of new or existing code.
- If we find some tooling, like runtime type checks, at least during testing (without big overheads, the test suite is already heavy), that would probably increase the utility of types and speed the adoption by core maintainers if they see the value.

TL;DR: Lets keep adopting typing in Conan codebase, but naturally/gradually, and not forcing it, not as a 2.0 task.

**Comment by luka-papez** [comment]
@memsharded that sounds great. I agree that adopting typing gradually is the way to go forward and that there are more important things to take care of. 

Sorry for not responding earlier, I missed the first e-mail and went to holiday mode after that.
